### ISSUE TITLE
[BUG]: In current architecture, copying textures to the encoder takes a really long time. It should not.

### ISSUE THREAD

**Issue by doctorpangloss** [Unity-Technologies/com.unity.webrtc#774]
### Package version

2.4.0-exp.8

### Environment

```markdown
* OS: macOS, Windows 11
* Unity version: Unity 2021.3.6f1
```

### Steps To Reproduce

<img width="2216" alt="Screen Shot 2022-07-15 at 11 51 34 AM" src="https://user-images.githubusercontent.com/2229300/179291923-121ced54-44d5-4d31-be7b-353616de6f97.png">

 1. On macOS, stream with `encoderImplementation=VideoToolbox`. This is to demonstrate that the underlying issue is not due to idiosyncrasies of implementing a VideoEncoder interface (as you did with NvEncoder). The accelerated video encoder on macOS comes from a libwebrtc authored factory.
 2. Does `OnFrameCaptured` really need to occur on the Render Thread? Why is a copy and NV12 conversion taking so long?
 3. Observe on VideoToolBox the actual encoder step is 1.29ms. Very short.

Is the actual encoding blocking the main thread? It is honestly hard to tell. Based on the profiler data, the last thing to happen before the Main Thread is unblocked is the UnityVideoEncoder.Encode step. This makes sense from the point of view of the code too.

As far as I understand, prior to 2.4.0-exp7 the NvEnc based encoder asked the graphics device to copy the frame into a buffer on the Render Thread, but not encode the frame.

I believe you have dealt with this copying issue a lot. In my experience I had the best latency in the 2.3... series, but I believe that shared the DirectX 12 texture backing the RenderTexture directly with NvEnc, which produced numerous artifacts as the engine wrote to the texture at the same time as the encoder thread read it. I do not 100% understand the underlying issues there.

I think it's essential to deal with this problem. Right now com.unity.webrtc is performing worse than NvFBC (i.e. NVIDIA Game Streaming) in part because of this architectural choice. It affects all architectures.

Profiler data:
[macOS Metal with custom rendering profiler data.data.zip](https://github.com/Unity-Technologies/com.unity.webrtc/files/9123270/macOS.Metal.with.custom.rendering.profiler.data.data.zip)

### Current Behavior

Poor multithreading of encoding.

### Expected Behavior

Encoding that is competitive with NvFBC's performance.

**Comment by doctorpangloss** [comment]
It looks like you are addressing this in https://github.com/Unity-Technologies/com.unity.webrtc/tree/experimental/direct-renderer . But why not pool RenderTextures from Unity side, and skip copying altogether?

**Comment by karasusan** [comment]
@mdodis
First we will try to solve the issue #803.

**Comment by karasusan** [comment]
Fixing each graphics device.
- DX11 #923
- DX12 #924 
- Metal #935
- GLCore/GLES #943

**Comment by vl4dimir** [comment]
Hi, I wanted to report a new bug but I see that this one seems to describe the same or similar issue, so I'll just post here. (I can also open a new ticket if that works better.)

We're seeing a significant performance drop on Android XR when the video stream starts, and profiling the code running on an Android device has led me to this:

<img width="1819" alt="Screenshot 2023-08-23 at 12 47 03" src="https://github.com/Unity-Technologies/com.unity.webrtc/assets/206409/79d750e8-dae3-465b-b882-24f6b185c644">

I haven't delved into the code too much but this mutex-locked section seems to be the main cause:

<img width="1137" alt="Screenshot 2023-08-23 at 12 44 49" src="https://github.com/Unity-Technologies/com.unity.webrtc/assets/206409/4b9a18f2-e68d-465b-8208-ac7943e69fd1">

So what I'd like to know is whether this performance issue be resolved by this ticket, or should I open another one? Seems related to what @doctorpangloss describes.

**Comment by karasusan** [comment]
@vl4dimir
Hi, we recognize that performance issue, and I am making a new draft PR #950 to fix it. Please review if you interested in.
This mutex is for controlling the encoding framerate, but if we sync the encoding framerate to application framerate, we don't need to use this mutex section.

**Comment by vl4dimir** [comment]
@karasusan I'm not sure if it's the mutex itself (waiting on it) that's causing the high CPU usage, or something within the `CaptureNextFrame` function. Is the capture being performed on the main thread?

(Looking at the PR now, thanks for sending.)

**Comment by karasusan** [comment]
@vl4dimir
In current implementation, the capture process is performed on the worker thread. The reason why the process should be called regularly. But there are a lag between the rendering process and the capturing process. This PR provides a option to developers to reduce the lag.

**Comment by karasusan** [comment]
#950 is merged.
